File: src/code/javascript/data_structures/doubly_linked_list.js
class ListNode {
    constructor(data) {
        this.data = data
        this.prev = null
        this.next = null
    }

    toString() {
        return `[${this.data}]`
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null
    }

    append(data) {
        if (!this.head) {
            this.head = new ListNode(data)
            return
        }

        let curr = this.head
        while (curr.next) {
            curr = curr.next
        }

        const new_node = new ListNode(data)
        curr.next = new_node
        new_node.prev = curr
    }

    delete(data) {
        if (!this.head) {
            return
        }

        if (this.head.data === data) {
            this.head = this.head.next
            if (this.head) {
                this.head.prev = null
            }
            return
        }

        let curr = this.head
        while (curr) {
            if (curr.data === data) {
                const prev_node = curr.prev
                prev_node.next = curr.next
                if (curr.next) {
                    curr.next.prev = prev_node
                }
                return
            }
            curr = curr.next
        }
    }

    reverse() {
        let curr = this.head
        let prev = null

        while (curr) {
            const nxt = curr.next
            curr.next = prev
            curr.prev = nxt
            prev = curr
            curr = nxt
        }

        this.head = prev
    }

    toString() {
        if (!this.head) {
            return 'None'
        }

        const nodes = []
        let curr = this.head

        while (curr) {
            nodes.push(curr.toString())
            curr = curr.next
        }

        return nodes.join(' <-> ') + ' <-> None'
    }
}

File: src/code/ruby/data_structures/doubly_linked_list.rb
class ListNode
  attr_accessor :data, :prev, :next

  def initialize(data)
    @data = data
    @prev = nil
    @next = nil
  end

  def to_s
    "[#{@data}]"
  end
end

class DoublyLinkedList
  attr_accessor :head

  def initialize
    @head = nil
  end

  def append(data)
    if @head.nil?
      @head = ListNode.new(data)
      return
    end

    curr = @head
    curr = curr.next while curr.next

    new_node = ListNode.new(data)
    curr.next = new_node
    new_node.prev = curr
  end

  def delete(data)
    return if @head.nil?

    if @head.data == data
      @head = @head.next
      @head.prev = nil if @head
      return
    end

    curr = @head
    while curr
      if curr.data == data
        prev_node = curr.prev
        prev_node.next = curr.next
        curr.next.prev = prev_node if curr.next
        return
      end
      curr = curr.next
    end
  end

  def reverse
    curr = @head
    prev = nil

    while curr
      nxt = curr.next
      curr.next = prev
      curr.prev = nxt
      prev = curr
      curr = nxt
    end

    @head = prev
  end

  def to_s
    return 'None' if @head.nil?

    nodes = []
    curr = @head
    while curr
      nodes << curr.to_s
      curr = curr.next
    end

    nodes.join(' <-> ') + ' <-> None'
  end
end
